core/blockchain.go

```diff
@@ -811,6 +811,14 @@ func (bc *BlockChain) ExportN(w io.Writer, first uint64, last uint64) error {
 //
 // Note, this function assumes that the `mu` mutex is held!
 func (bc *BlockChain) writeHeadBlock(block *types.Block) {
```

Comment:
There's also `core.HeaderChain.writeHeaders` where `rawdb.WriteCanonicalHash` is called. I'm just wondering if this could cause any issues. For example we may sync a header later than the stop block.

Reply:
writeHeaders checks something called procInterrupt(), which wraps a call to insertStopped(), which is injected when NewHeaderChain() is called. When we call StopInsert(), it makes procInterrupt() return true in writeHeaders before it updates the head header references etc. 
My understanding is that by calling bc.StopInsert() we prevent the HeaderChain code from continuing to update the head header.